src/channel/ws_0_8.ts

```diff
@@ -0,0 +1,564 @@
+   * expecting receiving message to contain same id
+   */
+  private sendId: number = 0;
+
+  /**
+   * subscriptions ids
+   * mapped by keys WSSubscriptions
+   */
+  readonly subscriptions: Map<string, number> = new Map();
+
+  /**
+   * Construct class and event listeners
+   * @param options WebSocketOptions
+   */
+  constructor(options: WebSocketOptions = {}) {
+    // provided existing websocket
+    const nodeUrl = options.nodeUrl || 'http://localhost:3000 '; // TODO: implement getDefaultNodeUrl default node when defined by providers?
+    this.nodeUrl = options.websocket ? options.websocket.url : nodeUrl;
+    this.websocket = options.websocket ? options.websocket : new WebSocket(nodeUrl);
+
+    this.websocket.addEventListener('open', this.onOpen.bind(this));
+    this.websocket.addEventListener('close', this.onCloseProxy.bind(this));
+    this.websocket.addEventListener('message', this.onMessageProxy.bind(this));
+    this.websocket.addEventListener('error', this.onError.bind(this));
+  }
+
+  private idResolver(id?: number) {
+    // unmanaged user set id
+    if (id) return id;
+    // managed id, intentional return old and than increment
+    // eslint-disable-next-line no-plusplus
+    return this.sendId++;
+  }
+
+  /**
+   * Send data over open ws connection
+   * * this would only send data on the line without awaiting 'response message'
+   * @example
+   * ```typescript
+   * const sentId = await this.send('starknet_method', params);
+   * ```
+   */
+  public send(method: string, params?: object, id?: number) {
+    if (!this.isConnected()) {
+      throw Error('WebSocketChannel.send() fail due to socket disconnected');
+    }
+    const usedId = this.idResolver(id);
+    const rpcRequestBody: JRPC.RequestBody = {
+      id: usedId,
+      jsonrpc: '2.0',
+      method,
+      ...(params && { params }),
+    };
+    // Stringify should remove undefined params
+    this.websocket.send(stringify(rpcRequestBody));
+    return usedId;
+  }
+
+  /**
+   * Send request and receive response over ws line
+   * This method abstract ws messages into request/response model
+   * @param method rpc method name
+   * @param params rpc method parameters
+   * @example
+   * ```typescript
+   * const response = await this.sendReceive('starknet_method', params);
+   * ```
+   */
+  public sendReceive(
+    method: string,
+    params?: {}
+  ): Promise<MessageEvent['data']['result'] | Error | Event> {
+    const sendId = this.send(method, params);
+
+    return new Promise((resolve, reject) => {
+      if (!this.websocket) return;
+      this.websocket.onmessage = ({ data }) => {
+        const message: JRPC.ResponseBody = JSON.parse(data);
+        if (message.id === sendId) {
+          if ('result' in message) {
+            resolve(message.result);
+          } else {
+            reject(Error(`error on ${method}, ${message.error}`));
+          }
+        }
+        // console.log(`data from ${method} response`, data);
```

Comment:
leftover?

Reply:
Still need it for testing as this PR can't be 100% complete before RPC 0.8 final release, Pathfinder WS Release and Devnet WS Release.
This is more like PrePR all working and can be checked but will be updated later.